myung6024/MyToyProject/app/src/main/java/com/runeanim/mytoyproject/ui/repo/RepoViewModel.kt

```diff
@@ -0,0 +1,100 @@
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.runeanim.mytoyproject.data.Result
+import com.runeanim.mytoyproject.data.source.local.entity.RepositoryEntity
+import com.runeanim.mytoyproject.data.source.remote.response.mapToPresentation
+import com.runeanim.mytoyproject.domain.GetRepositoriesUseCase
+import com.runeanim.mytoyproject.domain.RemoveAllRepositoriesUseCase
+import com.runeanim.mytoyproject.domain.SaveRepositoryUseCase
+import com.runeanim.mytoyproject.domain.SearchRepositoriesUseCase
+import com.runeanim.mytoyproject.util.Event
+import com.runeanim.mytoyproject.util.wrapEspressoIdlingResource
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class RepoViewModel(
+    private val searchRepositoriesUseCase: SearchRepositoriesUseCase,
+    private val saveRepositoriesUseCase: SaveRepositoryUseCase,
+    private val getRepositoriesUseCase: GetRepositoriesUseCase,
+    private val removeAllRepositoriesUseCase: RemoveAllRepositoriesUseCase
+) : ViewModel() {
+    private var isClickedRepoList: Boolean? = null
+
+    private val _items = MutableLiveData<List<RepositoryEntity>>(emptyList())
+    val items: LiveData<List<RepositoryEntity>> = _items
+
+    private val _openRepoEvent = MutableLiveData<Event<RepositoryEntity>>()
+    val openRepoEvent: LiveData<Event<RepositoryEntity>> = _openRepoEvent
+
+    private val _isLoading = MutableLiveData<Boolean>(false)
+    val isLoading: LiveData<Boolean> = _isLoading
+
+    private val _errorTextVisible = MutableLiveData<Boolean>(false)
+    val errorTextVisible: LiveData<Boolean> = _errorTextVisible
+
+    val searchRepositoryByKeyWord: (String) -> Unit
+        get() = fun(searchKeyWord: String) {
+            isClickedRepoList = false
+            _isLoading.value = true
+            _errorTextVisible.value = false
+            wrapEspressoIdlingResource {
+                viewModelScope.launch {
+                    val result = searchRepositoriesUseCase(searchKeyWord)
+                    if (result is Result.Success) {
+                        with(result.data.mapToPresentation()) {
+                            _items.value = this
+                            _errorTextVisible.value = size == 0
+                        }
+                    } else {
+                        _items.value = emptyList()
+                        _errorTextVisible.value = true
+                    }
+                    _isLoading.value = false
+                }
+            }
+        }
+
+    fun getClickedRepositoryHistory() {
+        isClickedRepoList = true
+        wrapEspressoIdlingResource {
+            viewModelScope.launch {
+                val result = getRepositoriesUseCase()
+                if (result is Result.Success) {
+                    _items.value = result.data
+                }
+            }
+        }
+    }
+
+    @VisibleForTesting
```

Comment:
이런 어노테이션도 있구나! 배워갑니다 👍

Reply:
`@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)`
이런식으로 넣을 수도 있는데 default가 private이라서 private땐 안써도 되긴 하는데 가독성을 위해서 넣는게 더 좋다고 하는 사람도 있긴 합니다.